core/src/avm1/globals/movie_clip.rs

```diff
@@ -1045,19 +1045,29 @@ fn set_mask<'gc>(
     activation: &mut Activation<'_, 'gc>,
     args: &[Value<'gc>],
 ) -> Result<Value<'gc>, Error<'gc>> {
-    let mask = args
-        .get(0)
-        .unwrap_or(&Value::Undefined)
-        .coerce_to_object(activation)
-        .as_display_object();
+    let mask = if let Some(m) = args.get(0) {
+        match m {
+            Value::Undefined | Value::Null => None,
+            _ => {
+                let start_clip = activation.target_clip_or_root();
+                let mask = activation.resolve_target_display_object(start_clip, *m, false)?;
+                if mask.is_none() {
+                    return Ok(Value::Bool(false));
+                }
+                mask
+            }
+        }
+    } else {
+        return Ok(Value::Undefined);
+    };
```

Comment:
One more style nit. This might be cleaner with a single `match` instead of both `if` and `match`:
```
let mask = match args.get(0) {
   None => return Ok(Value::Undefined),
  Some(Value::Undefined | Value::Null) => None,
  Some(m) => // ... etc
}
```